### ISSUE TITLE
DATETIME2 conversion exception

### ISSUE THREAD

**Issue by nico525** [pymssql/pymssql#695]
Hi guys,
I am using FreeTDS 1.2.18 on an apple M1 and pymssql 2.1.5 in combination with sqlalchemy.
In general, everything seems to work except if I include the valid_to column that is generate always as row end for my system versioned table.
I am getting an ValueError("hour must be in 0..23") that is probably due to some type checks:
`return self.cursor.fetchall()
  File "src/pymssql.pyx", line 549, in pymssql.Cursor.fetchall
  File "src/pymssql.pyx", line 500, in pymssql.Cursor.getrow
  File "src/_mssql.pyx", line 449, in _mssql.MSSQLRowIterator.__next__
  File "src/_mssql.pyx", line 1196, in _mssql.MSSQLConnection.fetch_next_row
  File "src/_mssql.pyx", line 1367, in _mssql.MSSQLConnection.get_row
  File "src/_mssql.pyx", line 856, in _mssql.MSSQLConnection.convert_db_value
ValueError: hour must be in 0..23`

row end columns usually are DATETIME2 and their default date is 9999-12-31 23:59:59.9999999
Selecting all other DATETIME 2 values work fine.

minimal example:

> CREATE TABLE dbo.test (
> valid_from DATETIME2 GENERATED ALWAYS AS ROW START NOT NULL,
> valid_to DATETIME2 GENERATED ALWAYS AS ROW END NOT NULL,
> PERIOD FOR SYSTEM_TIME (valid_from,valid_to),
> id INTEGER NOT NULL IDENTITY(1,1) PRIMARY KEY, 
> test VARCHAR(255) NOT NULL, 
> )
> WITH (SYSTEM_VERSIONING = ON (HISTORY_TABLE = dbo.test_history));
> 
> INSERT INTO dbo.test(test)VALUES('TEST');
> 
> SELECT valid_from, valid_to, id, test FROM dbo.test

**Comment by termim** [comment]
Please provide pymssql/FreeTDS versions and a sample query that demonstrates the error.

**Comment by nico525** [comment]
I updated my original post to have a minimal example as well as the version

**Comment by Ahmed1Ossama13** [comment]
Any update on this issue?
Seems like not just related to ROW END columns, but any value in a datetime2 that is like '23:59:59.999' will trigger this issue.

Any workaround?

**Comment by fivejjs** [comment]
same issue here.

**Comment by pauliev** [comment]
Getting same issue as well
```
File "src/pymssql/_pymssql.pyx", line 521, in pymssql._pymssql.Cursor.fetchone
  File "src/pymssql/_pymssql.pyx", line 511, in pymssql._pymssql.Cursor.getrow
  File "src/pymssql/_mssql.pyx", line 470, in pymssql._mssql.MSSQLRowIterator.__next__
  File "src/pymssql/_mssql.pyx", line 1213, in pymssql._mssql.MSSQLConnection.fetch_next_row
  File "src/pymssql/_mssql.pyx", line 1384, in pymssql._mssql.MSSQLConnection.get_row
  File "src/pymssql/_mssql.pyx", line 875, in pymssql._mssql.MSSQLConnection.convert_db_value
```

**Comment by pauliev** [comment]
Alright... I am not positive... my issue is hitting the datetime2 type in SQL Server. I noticed the following in `_mssql.pyx` starting around line 871

```
        elif dbtype in (SQLDATETIM4, SQLDATETIME2):
            dbconvert(self.dbproc, dbtype, data, -1, SQLDATETIME,  <--- HERE
                <BYTE *>&dt, -1)
            dbdatecrack(self.dbproc, &di, <DBDATETIME *><BYTE *>&dt)
            return datetime.datetime(di.year, di.month, di.day,
                di.hour, di.minute, di.second, di.millisecond * 1000)
```

The value `SQLDATETIME` is mapped to `SYBDATETIME` should this not be `SQLDATETIME2`

**Comment by cdquintero97** [comment]
facing the same issue.

**Comment by visch** [comment]
Another way to replicate this

```

CREATE TABLE example (
	CreateDateUtc datetime2 NOT NULL,
	UpdateDateUtc datetime2 NOT NULL
);

INSERT INTO example

( CreateDateUtc, UpdateDateUtc)

VALUES('2021-01-11 23:59:59.9989862', '2021-01-11 23:59:59.9989862')
```

Select from this table with `pymssql`

**Comment by justinzw2xep** [comment]
I am experiencing the same issue, is there plan for fixing; or a workaround?

**Comment by termim** [comment]
Closed in bc2cc5040ba0453bd22798bd49ed4d39ed8df3e3
